graph/src/components/store/entity_cache.rs

```diff
@@ -209,11 +209,59 @@ impl EntityCache {
             id_is_bytes: id_is_bytes,
         };
 
+        // Get the entities from the store and insert it into the cache if it's not already there
         let entities = self.store.get_derived(&query)?;
-        entities.iter().for_each(|(key, e)| {
-            self.current.insert(key.clone(), Some(e.clone()));
+
+        entities.iter().for_each(|(k, v)| {
+            // Only insert to the cache if it's not already there
+            if !self.current.contains_key(k) {
+                self.current.insert(k.clone(), Some(v.clone()));
+            }
+        });
+
+        // Insert `None` for entities that are not in the store and not in the current cache
+        // but are present in updates or handler_updates
+        // The updates will be applied before returning the entities
+        self.updates.iter().for_each(|(k, _)| {
+            if !self.current.contains_key(k) {
+                self.current.insert(k.clone(), None);
+            }
         });
-        let entities: Vec<Entity> = entities.values().cloned().collect();
+
+        self.handler_updates.iter().for_each(|(k, _)| {
+            if !self.current.contains_key(k) {
+                self.current.insert(k.clone(), None);
+            }
+        });
```

Comment:
Actually, I think I understand why you are inserting the `None` here - so that you can drive the lookup loop below by iterating over `self.current`. Wouldn't it be better to iterate over `self.updates` and `self.handler_updates` and merge whatever is found there into the entities from the store? That way, you don't need to iterate over `self.current` and therefore don't need to insert `None`

Reply:
Ok, i was intially unaware of updates and handlerUpdates so thought everything thats updated is in self.current. Good to know. Will make those changes

Reply:
On second thoughts i think this is actually the best approach i could think of, its something similar that happens in `store.get` besides that if i were to iterate over `self.updates` and `handler_updates` i would still need to create a `None` entity and then call EntityOp.apply_to() on the None entity in the case where the entities are newly created entities within the block / handler.

Reply:
Same thing happens with store.get() if the entity is not present in the store a `None` value is getting inserted into `self.current` [here](https://github.com/graphprotocol/graph-node/blob/d97ca7c6e326f3f4eb032794fd81335006641c17/graph/src/components/store/entity_cache.rs#L172)

Then later entity operations is being applied to this `None` value [here](https://github.com/graphprotocol/graph-node/blob/d97ca7c6e326f3f4eb032794fd81335006641c17/graph/src/components/store/entity_cache.rs#L188)

Reply:
My point is that we don't really know here whether the entity is in the store or not; and it seems more logical to me to compose the list of entities we return of entities from the store + what we find in `self.updates` and `self.handler_updates`.

Reply:
Ok, i see your point now. That was an oversight from my part thanks for pointing. Since we insert all the entities from `store.get_derived()` to  `self.current` i thought we can be sure that if its not in `self.current` its not both in the store and cache, oversaw the part where updates could contain entities that are not related to the ones we are searching for

Reply:
I think we still would need to iterate over `self.current` since when we move on the next block and a new blockstate is created `updates` becomes empty, only `self.current` now has the updated entities, neither store nor `updates` has them